### ISSUE TITLE
Bugs on comments ending with Chinese characters followed by a return statement

### ISSUE THREAD

**Issue by cyh2004** [rust-lang/rust-clippy#12491]
### Summary

```rust
fn main() {
	if (true) {
		// anything一些中文
        return;
	}
}
```

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
      Checking test-bug v0.1.0 (/home/chen/test-bug)
warning: unnecessary parentheses around `if` condition
 --> src/main.rs:2:5
  |
2 |     if (true) {
  |        ^    ^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
2 -     if (true) {
2 +     if true {
  |

thread 'rustc' panicked at compiler/rustc_span/src/lib.rs:1787:17:
assertion failed: bpos.to_u32() >= mbc.pos.to_u32() + mbc.bytes as u32
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:144:5
   3: <rustc_span::source_map::SourceMap>::lookup_char_pos
   4: <rustc_errors::emitter::FileWithAnnotatedLines>::collect_annotations
   5: <rustc_errors::emitter::EmitterWriter>::emit_message_default::{closure#0}
   6: <rustc_errors::emitter::EmitterWriter>::emit_messages_default
   7: <rustc_errors::emitter::EmitterWriter as rustc_errors::emitter::Emitter>::emit_diagnostic
   8: <rustc_errors::json::Diagnostic>::from_errors_diagnostic
   9: <rustc_errors::json::JsonEmitter as rustc_errors::emitter::Emitter>::emit_diagnostic
  10: <rustc_errors::DiagCtxtInner>::emit_diagnostic_without_consuming::{closure#2}
  11: rustc_interface::callbacks::track_diagnostic
  12: <rustc_errors::DiagCtxtInner>::emit_diagnostic_without_consuming
  13: <() as rustc_errors::diagnostic_builder::EmissionGuarantee>::diagnostic_builder_emit_producing_guarantee
  14: rustc_middle::lint::struct_lint_level::struct_lint_level_impl::<alloc::string::String>
  15: clippy_lints::returns::check_final_expr
  16: clippy_lints::returns::check_block_return
  17: clippy_lints::returns::check_final_expr
  18: clippy_lints::returns::check_block_return
  19: <clippy_lints::returns::Return as rustc_lint::passes::LateLintPass>::check_fn
  20: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_fn
  21: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_item
  22: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_mod
  23: rustc_lint::late::check_crate
  24: rustc_interface::passes::analysis
      [... omitted 1 frame ...]
  25: rustc_interface::interface::run_compiler::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#0}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: rustc 1.76.0 (07dca489a 2024-02-04) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.76 (07dca48 2024-02-04)

warning: `test-bug` (bin "test-bug") generated 1 warning (run `cargo clippy --fix --bin "test-bug"` to apply 1 suggestion)
error: could not compile `test-bug` (bin "test-bug"); 1 warning emitted
  ```

  </p>
</details>

**Comment by Lethe10137** [comment]
Reproductable in a different environment. Both stable and nightly.
```
rustc -Vv
rustc 1.78.0-nightly (ef324565d 2024-02-27)
binary: rustc
commit-hash: ef324565d071c6d7e2477a195648549e33d6a465
commit-date: 2024-02-27
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

```
rustc -Vv
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```

**Comment by y21** [comment]
This looks very much like the same ICE that was reported in #12328, but with a repro this time. Thanks for reporting it!

**Comment by yhx-12243** [comment]
The problem is most likely to come from the following snippets.
https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/returns.rs#L466-475

**Comment by flip1995** [comment]
`BytePos` should really be marked `unsafe`
